README.md
# gferplot
[![Travis-CI Build Status](https://travis-ci.org/Yuanchao-Xu/gferplot.svg?branch=master)](https://travis-ci.org/Yuanchao-Xu/gferplot)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/gferplot)](https://cran.r-project.org/package=gferplot)


## Installation

Released version from CRAN, for beginners and normal users:

```R
install.packages("gferplot")
```

Development version from github, for experienced users and those who are interested in investigating:

```R
install.packages("devtools")
# You can ignore the line above, if you have already installed devtools
devtools::install_github("Yuanchao-Xu/gferplot")
```

**Official Website is [http://yuanchao-xu.github.io/gferplot](https://yuanchao-xu.github.io/gferplot)**

gferplot is an R package, designed for green finance and environmental risk research plotting

**If you feel gferplot is of a little help, please cite it as following:**

Yuanchao Xu (2017). gferplot: Green Finance and Environmental Risk Plotting. R package version 0.1.7.
  https://CRAN.R-project.org/package=gferplot






NEWS
gfer 0.1.1
==========
Date: 2017-8-8

- add waternomic plot and circular plot










R/plotCircular.R

#' plotScatterPie
#'
#' plot scatter pie chart for multidimension analysis, such as waternomics. This plot can
#' provide information about water use/wastewater of each provinces and GDP mix of each provinces,
#' see examples.
#' @param data a dataframe showing different management intersections. See the data frame in the example
#' @param t is transpose the dataframe, by default, lines flow from row to column, if t == TRUE, lines will
#' flow from columns to rows. Once transposed,
#' @param ifsep if separate row and col categories in the chart, default is TRUE
#' @param trans transparency of the chart's lines, default is 0.3
#' @param highlight a string or string array of highlighted items, MUST be selected from first column (which represents names) or colnames.
#' if highlight has more than 2 items, they should belong to same category, either colnames, or names. One name and one column name is not allowed.
#' @param xlim x limit of the chart, default is c(-1, 1)
#' @param ylim y limte of the chart, default is c(-1, 1)
#' @import circlize
#' @examples
#'
#'
#'
#' \dontrun{
#' plotChord(cm)
#' plotChord(cm, t = T)
#' plotChord(cm, highlight = 'MEP')
#' plotChord(cm, highlight = 'Investment')
#' }
#'
plotChord <- function(data, t = FALSE, ifsep = TRUE, trans = 0.3, highlight = NULL, xlim = c(-1, 1), ylim = c(-1, 1)) {

  # tidy up data, change data frame to matrix with dimension names filled
  m <- as.matrix(data[2:ncol(data)])
  dimnames(m)[[1]] <- data[,1]

  if (t == TRUE) m <- t(m)
  # clean the canvass
  circos.clear()


  gaps <- NULL
  # set gaps
  if (ifsep == TRUE) {
    # set up small gap and big gap, this is to seperate the whole circle into up and down part
    small_gap = 1
    big_gap = 20

    # calculate how many sectors involved n_sector
    nr = nrow(m)
    nc = ncol(m)
    n_sector = nr + nc

    # calculate each sector takes up how many width, 0.5 means seperate into 2 parts, 0.25 means 4 parts
    row_sector_degree = (360 - small_gap*(n_sector - 2) - big_gap*2) * 0.5 + small_gap*(nr-1)

    start_degree = 0 - (180 - row_sector_degree)/2

    # calculate starting degree
    gaps = c(rep(small_gap, nrow(m) - 1), big_gap, rep(small_gap, ncol(m) - 1), big_gap)
  }




  col <- NULL
  # set up color, if highlight is selected
  if (!is.null(highlight)) {
    col <- m
    col[] <- 'grey'


    if (highlight %in% dimnames(m)[[1]]) col[highlight,] <- 'red'
    if (highlight %in% dimnames(m)[[2]]) col[,highlight] <- 'red'
  }

  # parameters

  circos.par(gap.after = gaps, start.degree = start_degree, canvas.xlim = xlim, canvas.ylim = ylim,
             points.overflow.warning = FALSE)

  # plot basic plot
  chordDiagram(m, annotationTrack = "grid", transparency = trans, col = col)

  # change label settings
  circos.track(track.index = 1, panel.fun = function(x, y) {
    circos.text(CELL_META$xcenter, CELL_META$ylim[1] + 2, CELL_META$sector.index,
                facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5))
  }, bg.border = NA)
  abline(h = 0, lty = 2, col = "#00000080")


}





R/plotWaternomics.R


# this is for setting plot radius
getRadius <- function(y, small = 1.5, medium = 3, large = 5) {
  r <- sapply(y, function(x) {
    if (x <= 40000 & x >= 15000) {
      x <- medium
    } else if (x < 15000) {
      x <- small
    } else {
      x <- large
    }
  })
  return (r)
}


#' plotScatterPie
#'
#' plot scatter pie chart for multidimension analysis, such as waternomics. This plot can
#' provide information about water use/wastewater of each provinces and GDP mix of each provinces,
#' see examples.
#' @param data a dataframe with information like x, y, r, etc. Refer to \code{GDPmix}
#' @param labelLine how far is label to pie chart, can be left with default value.
#' @importFrom graphics plot abline text
#' @import ggplot2
#' @importFrom scatterpie geom_scatterpie

plotScatterPlot <- function(data, labelLine = NULL) {

  with (data, {
    layer_basic <- ggplot()
    #  layer_points <- geom_point(data = data, aes(x, y, size = radius))

    data$radius <- getRadius(data$GDP)


    layer_pie <- geom_scatterpie(data = data, aes(x, y, r = radius),
                                 cols = colnames(data)[3:5], color = NA)

    if (is.null(labelLine)) labelLine <- max(data$radius)/3

    layer_label <- geom_text_repel(data = data, aes(x, y, label = Province),
                                   point.padding = unit(labelLine, "lines"))

    layer_lines <-

      #layer_legend <- geom_scatterpie_legend(data$radius, x= 0, y=0)

      layer_plot <- layer_basic + layer_pie + layer_label + coord_equal()


    style <- ggstyle()



    print(layer_plot + style)


  })


}


#' @import ggplot2
ggstyle <- function() {
  a <- theme_classic() +
    theme(axis.line = element_line(size = 1, colour = "black"))
             #panel.background = element_rect(fill = "white"),
             #panel.grid.major = element_line(colour = "grey50"))
}

########################################################################################
#######################################################################################
#'
#'
#'
#' #' plotWaternomics
#' #'
#' #' plot special waternomics chart for CWR
#' #'
#' #'
#' #' @param data a dataframe containing columns of GDP, Value-added of fisrt/second/third industry, x and y
#' #' x and y are the coordinates of the plot, and it varies depending on needs, can be per capita
#' #' water use vs. per capita wastewater, etc. See example of year 2015 by \code{GDPmix}
#' #'
#' #' Columns should be exactly the same as GDPmix, which means, first column is province, seconnd is x,
#' #' third is y, fourth is Frist, etc. Load GDPmix to have a look at the columns.
#' #' @param title chart title
#' #' @param xname x axis name
#' #' @param yname y axis name
#' #' @param small radius of small circle, default is 1.5
#' #' @param medium radius of medium circle, default is 3
#' #' @param large radius of large circle, default is 5
#' #' @param legend whether to show legend, default is TRUE
#' #' @param label whether to show label, default is TRUE
#' #' @param xmean a line showing mean value of x
#' #' @param ymean a line showing mean value of y
#' #' @param line whether to show ablines, default is TRUE
#' #' @importFrom mapplots add.pie
#' #' @importFrom graphics plot plot.new
#' #' @importFrom grDevices rgb
#' #' @importFrom plotrix thigmophobe.labels
#' #' @export
#' #' @examples
#' #' \dontrun{
#' #' data(GDPmix)
#' #' plotwaternomics(GDPmix)
#' #' }
#'
#'
#'
#' plotWaternomics <- function(data, title = '', xname = '', yname = '', small = 1.5, medium = 3, large = 5,
#'                             legend = TRUE, label = TRUE, xmean = NULL, ymean = NULL, line = TRUE) {
#'   # plot basic plot based on x and y
#'   #plot.new()
#'
#'   # for some buffer space, extend the x and y axis, redefine them first
#'   xlim <- getLim(data$x)
#'
#'   ylim <- getLim(data$y)
#'
#'
#'   plot(data$x, data$y, main = title, xlab = xname, ylab = yname,
#'        xlim = xlim, ylim = ylim,
#'        bty = 'n')
#'
#'
#'
#'   provinceNum <- nrow(data)
#'
#'   # decide radius of the pie charts
#'   r <- getRadius(data$GDP, small, medium, large)
#'
#'   # define CWR color
#'   col <- c(rgb(107/255, 128/255, 51/255), rgb(2/255, 3/255, 3/255), rgb(13/255, 119/255, 185/255))
#'
#'
#'
#'
#'   # add pie charts
#'   for (i in 1:provinceNum) {
#'
#'     add.pie(x = data$x[i], y = data$y[i], z = c(data$First[i], data$Second[i], data$Third[i]),
#'             labels = '', radius = r[i], col = col, border = "white")
#'     #thigmophobe.labels(data$x[i], data$y[i], data$Province[i], offset = r[i]/3)
#'
#'   }
#'   if (legend == TRUE) {
#'     legend('topright', c("Agriculture","Industry","Services"), cex = 0.8,
#'            fill = col)
#'   }
#'   if (label == TRUE) {
#'     thigmophobe.labels(data$x, data$y, data$Province, offset = 1)
#'   }
#'
#'
#'   if (line == TRUE) {
#'     # add ablines of national mean or mean
#'     if (is.null(xmean)) xmean <- round(mean(data$x), 2)
#'     if (is.null(ymean)) ymean <- round(mean(data$y), 2)
#'     abline(h = ymean, v = xmean, col = '#c00000', lwd = 3, lty = 2)
#'   }
#'
#'   #spread.labels(data$x, data$y, data$Province, offsets = 0.01)
#'
#' }
#'
#'
#'
#' #' plotWaternomics_legend
#' #'
#' #' legend for plot special waternomics chart for CWR
#' #'
#' #' @param label whether to show label, default is TRUE
#' #' @importFrom mapplots add.pie
#' #' @importFrom graphics plot abline text
#' #' @importFrom grDevices rgb
#' #' @export
#' #' @examples
#' #' \dontrun{
#' #' plotwaternomics_legend()
#' #' }
#' plotWaternomics_legend <- function(label = TRUE) {
#'   r <- c(5, 3, 1.5)
#'   x <- c(0, 0, 0)
#'   y <- c(7, 16, 21)
#'   plot (x, y, xlim = c(-3, 3), ylim = c(0, 23), xaxt = 'n', yaxt = 'n', bty = 'n')
#'
#'   col <- c(rgb(107/255, 128/255, 51/255), rgb(2/255, 3/255, 3/255), rgb(13/255, 119/255, 185/255))
#'
#'   for (i in 1:3) {
#'     add.pie(x = x[i], y = y[i], z = c(1, 1, 1), radius = r[i], labels = '', col = col, border = 'white')
#'   }
#'   if (label == TRUE) {
#'     size <- 1.2
#'     text(-0.7, 8.5, "Service", col = 'white', cex = size)
#'     text(0.7, 8.5, "Agriculture", col = 'white', cex = size)
#'     text(0, 4, "Industry", col = 'white', cex = size)
#'   }
#' }
#'
#'
#'
#'
#'
#' getLim <- function(x) {
#'   # decide axis interval
#'   dig <- nchar(min(round(x)))
#'
#'   x1 <- round(min(x) - (max(x) - min(x))/length(x), -(dig - 1))
#'   x2 <- round(max(x) + (max(x) - min(x))/length(x), -(dig - 1))
#'
#'   return(c(x1, x2))
#' }
#'
#'
#'
#'


R/plotScatterPie.R


# this is for setting plot radius
getRadius <- function(y, small = 1.5, medium = 3, large = 5) {
  r <- sapply(y, function(x) {
    if (x <= 40000 & x >= 15000) {
      x <- medium
    } else if (x < 15000) {
      x <- small
    } else {
      x <- large
    }
  })
  return (r)
}


#' plotScatterPie
#'
#' plot scatter pie chart for multidimension analysis, such as waternomics. This plot can
#' provide information about water use/wastewater of each provinces and GDP mix of each provinces,
#' see examples.
#' @param data a dataframe with information like x, y, r, label. See examples about how to assign these columns as required.
#' @param pieRange define which column to which column to be presented by pie chart, see examples
#' @param pieColor color for different colors in pie chart
#' @param labelLine how far is label to pie chart, can be left with default value.
#' @param xmeanLine if plot x mean line
#' @param ymeanLine if plot y mean line
#' @importFrom graphics plot abline text
#' @importFrom ggrepel geom_text_repel
#' @importFrom scatterpie geom_scatterpie
#' @import ggplot2
#' @examples
#'
#'
#' GDPColor_CWR <- c("#6B8033", "#020303", "#0D77B9")
#'
#' # Change colnames so that the function can recognize x, y, r, label
#'
#' colnames(GDPmix) <- c('label', 'r', '1st', '2nd', '3rd', 'x', 'y')
#'
#' \dontrun{
#' plotScatterPie(GDPmix, pieRange = 3:5, pieColor = GDPColor_CWR)
#' }
#'
#'

plotScatterPie <- function(data, pieRange, pieColor = NULL, xmeanLine = TRUE, ymeanLine = TRUE, labelLine = NULL) {

  ## input check
  if(is.null(pieRange)) stop("You have to assign which column to which column to be presented by pie chart.")
  if(length(pieRange)!=length(pieColor)) stop("Length of pieRange and pieColor should be the same.")


  with (data, {
    layer_basic <- ggplot()
    #  layer_points <- geom_point(data = data, aes(x, y, size = radius))

    data$radius <- getRadius(data$r)


    layer_pie <- geom_scatterpie(data = data, aes(x, y, r = radius),
                                 cols = colnames(data)[pieRange], color = NA)

    if (is.null(labelLine)) labelLine <- max(data$radius)/3

    layer_label <- geom_text_repel(data = data, aes(x, y, label = label),
                                   point.padding = unit(labelLine, "lines"))
    #layer_legend <- geom_scatterpie_legend(data$radius, x= 0, y=0)
    if (xmeanLine == TRUE) layer_basic <- layer_basic + geom_vline(xintercept = mean(data$x), color = 'red', size = 1.5, linetype = 2)
    if (ymeanLine == TRUE) layer_basic <- layer_basic + geom_hline(yintercept = mean(data$y), color = 'red', size = 1.5, linetype = 2)


    layer_plot <- layer_basic + layer_pie + layer_label + scale_fill_manual(values=pieColor) +
      coord_equal() + ggstyle()

    # # set up xlim and ylim to keep the chart ratio more or less 4:3
    # xlim <- getLim(data$x)
    # ylim <- getLim(data$y)
    #
    # xlim <- adjustxlim(xlim, ylim)
    #
    # layer_plot <- layer_plot + xlim(xlim) + ylim(ylim)

    print(layer_plot)

  })
}


#' @import ggplot2
ggstyle <- function() {
  a <- theme_set(theme_bw())
  #   theme(axis.line = element_line(size = 1, colour = "black"))
             #panel.background = element_rect(fill = "white"),
             #panel.grid.major = element_line(colour = "grey50"))
  return(a)
}

adjustxlim <- function(xlim, ylim, ratio = 0.75) {
  length_y <- max(ylim) - min(ylim)
  length_x <- max(xlim) - min(xlim)
  if (length_x < length_y/ratio) xlim <- c(min(xlim), min(xlim) + length_y)
  return(xlim)
}

getLim <- function(x) {
  # decide axis interval
  dig <- nchar(min(round(x)))

  x1 <- round(min(x) - (max(x) - min(x))/length(x), -(dig - 1))
  x2 <- round(max(x) + (max(x) - min(x))/length(x), -(dig - 1))

  return(c(x1, x2))
}

########################################################################################
#######################################################################################
#'
#'
#'
#' #' plotWaternomics
#' #'
#' #' plot special waternomics chart for CWR
#' #'
#' #'
#' #' @param data a dataframe containing columns of GDP, Value-added of fisrt/second/third industry, x and y
#' #' x and y are the coordinates of the plot, and it varies depending on needs, can be per capita
#' #' water use vs. per capita wastewater, etc. See example of year 2015 by \code{GDPmix}
#' #'
#' #' Columns should be exactly the same as GDPmix, which means, first column is province, seconnd is x,
#' #' third is y, fourth is Frist, etc. Load GDPmix to have a look at the columns.
#' #' @param title chart title
#' #' @param xname x axis name
#' #' @param yname y axis name
#' #' @param small radius of small circle, default is 1.5
#' #' @param medium radius of medium circle, default is 3
#' #' @param large radius of large circle, default is 5
#' #' @param legend whether to show legend, default is TRUE
#' #' @param label whether to show label, default is TRUE
#' #' @param xmean a line showing mean value of x
#' #' @param ymean a line showing mean value of y
#' #' @param line whether to show ablines, default is TRUE
#' #' @importFrom mapplots add.pie
#' #' @importFrom graphics plot plot.new
#' #' @importFrom grDevices rgb
#' #' @importFrom plotrix thigmophobe.labels
#' #' @export
#' #' @examples
#' #' \dontrun{
#' #' data(GDPmix)
#' #' plotwaternomics(GDPmix)
#' #' }
#'
#'
#'
#' plotWaternomics <- function(data, title = '', xname = '', yname = '', small = 1.5, medium = 3, large = 5,
#'                             legend = TRUE, label = TRUE, xmean = NULL, ymean = NULL, line = TRUE) {
#'   # plot basic plot based on x and y
#'   #plot.new()
#'
#'   # for some buffer space, extend the x and y axis, redefine them first
#'   xlim <- getLim(data$x)
#'
#'   ylim <- getLim(data$y)
#'
#'
#'   plot(data$x, data$y, main = title, xlab = xname, ylab = yname,
#'        xlim = xlim, ylim = ylim,
#'        bty = 'n')
#'
#'
#'
#'   provinceNum <- nrow(data)
#'
#'   # decide radius of the pie charts
#'   r <- getRadius(data$GDP, small, medium, large)
#'
#'   # define CWR color
#'   col <- c(rgb(107/255, 128/255, 51/255), rgb(2/255, 3/255, 3/255), rgb(13/255, 119/255, 185/255))
#'
#'
#'
#'
#'   # add pie charts
#'   for (i in 1:provinceNum) {
#'
#'     add.pie(x = data$x[i], y = data$y[i], z = c(data$First[i], data$Second[i], data$Third[i]),
#'             labels = '', radius = r[i], col = col, border = "white")
#'     #thigmophobe.labels(data$x[i], data$y[i], data$Province[i], offset = r[i]/3)
#'
#'   }
#'   if (legend == TRUE) {
#'     legend('topright', c("Agriculture","Industry","Services"), cex = 0.8,
#'            fill = col)
#'   }
#'   if (label == TRUE) {
#'     thigmophobe.labels(data$x, data$y, data$Province, offset = 1)
#'   }
#'
#'
#'   if (line == TRUE) {
#'     # add ablines of national mean or mean
#'     if (is.null(xmean)) xmean <- round(mean(data$x), 2)
#'     if (is.null(ymean)) ymean <- round(mean(data$y), 2)
#'     abline(h = ymean, v = xmean, col = '#c00000', lwd = 3, lty = 2)
#'   }
#'
#'   #spread.labels(data$x, data$y, data$Province, offsets = 0.01)
#'
#' }
#'
#'
#'
#' #' plotWaternomics_legend
#' #'
#' #' legend for plot special waternomics chart for CWR
#' #'
#' #' @param label whether to show label, default is TRUE
#' #' @importFrom mapplots add.pie
#' #' @importFrom graphics plot abline text
#' #' @importFrom grDevices rgb
#' #' @export
#' #' @examples
#' #' \dontrun{
#' #' plotwaternomics_legend()
#' #' }
#' plotWaternomics_legend <- function(label = TRUE) {
#'   r <- c(5, 3, 1.5)
#'   x <- c(0, 0, 0)
#'   y <- c(7, 16, 21)
#'   plot (x, y, xlim = c(-3, 3), ylim = c(0, 23), xaxt = 'n', yaxt = 'n', bty = 'n')
#'
#'   col <- c(rgb(107/255, 128/255, 51/255), rgb(2/255, 3/255, 3/255), rgb(13/255, 119/255, 185/255))
#'
#'   for (i in 1:3) {
#'     add.pie(x = x[i], y = y[i], z = c(1, 1, 1), radius = r[i], labels = '', col = col, border = 'white')
#'   }
#'   if (label == TRUE) {
#'     size <- 1.2
#'     text(-0.7, 8.5, "Service", col = 'white', cex = size)
#'     text(0.7, 8.5, "Agriculture", col = 'white', cex = size)
#'     text(0, 4, "Industry", col = 'white', cex = size)
#'   }
#' }
#'
#'
#'
#'
#'
#'
#'
#'

